indico/modules/users/tasks.py

```diff
@@ -37,3 +60,211 @@ def update_gravatars(user=None):
+
+
+def export_user_data(user, options):
+    """Generate a zip file with all user data and save it in user.data_export_request."""
+    export_request = user.data_export_request
+    if export_request and export_request.is_running:
+        return
+
+    # ughh why isn't there an ORM upsert..
+    if export_request:
+        db.session.delete(export_request)
+        db.session.commit()
+    export_request = DataExportRequest(user=user, selected_options=options, state=DataExportRequestState.running)
+    db.session.commit()
+
+    temp_file = NamedTemporaryFile(suffix='.zip', dir=config.TEMP_DIR, delete=False)
+    try:
+        _export_user_data(export_request, temp_file)
+    except Exception:
+        logger.exception('Could not create a user data export %r', export_request)
+        export_request.fail()
+        db.session.commit()
+        notify_data_export_failure(export_request)
+    finally:
+        Path(temp_file.name).unlink(missing_ok=True)
+
+
+def _export_user_data(export_request, buffer):
+    generate_zip(export_request, buffer, config.MAX_DATA_EXPORT_SIZE * 1024**2)
+    buffer.seek(0)
+
+    file = File(filename='data-export.zip', content_type='application/zip')
+    try:
+        file.save(('user', export_request.user.id), buffer)
+        file.claim()
+        export_request.succeed(file)
+        db.session.commit()
+    except Exception as exc:
+        with suppress(Exception):
+            file.delete()
+        raise exc
+    else:
+        notify_data_export_success(export_request)
+
+
+@celery.task(name='export_user_data')
+def export_user_data_task(user, options):
+    export_user_data(user, options)
+
+
+def get_data(export_request):
+    from indico.modules.users.schemas import UserDataExportSchema
+
+    user = export_request.user
+    options = export_request.selected_options
+
+    options_map = {
+        DataExportOptions.contribs: ('contributions', 'subcontributions'),
+        DataExportOptions.abstracts_papers: ('abstracts', 'papers'),
+        DataExportOptions.misc: ('miscellaneous',),
+    }
+    fields = []
+    for opt in options:
+        fields += options_map.get(opt, (opt.name,))
+    return UserDataExportSchema(only=fields).dump(user)
+
+
+def generate_zip(export_request, temp_file, max_size):
+    data = get_data(export_request)
+    with ZipFile(temp_file, 'w', allowZip64=True) as zip_file:
+        for key, subdata in data.items():
+            zip_file.writestr(f'{key}.yaml', yaml.dump(subdata))
+        size = 0
+        for file in collect_files(export_request):
+            size += getattr(file, 'file', file).size
+            if size <= max_size:
+                write_file(zip_file, file)
+            else:
+                export_request.max_size_exceeded = True
+                logger.warning('User data export exceeds size limit, exporting only partial data: %r',
+                               export_request)
+                break
+
+
+def collect_files(export_request):
```

Comment:
I didn't check how but this yields duplicate files (which are then added to the zip file twice, with the exact same path) - I only noticed this because when unpacking the zip file with Total Commander into an empty folder it asked me whether to overwrite some files.

Reply:
Looks like it's files in editables. So I think deduplicating the list returned from this function is sufficient

Reply:
Probably same file referenced from multiple revisions. Including the revision id in the path should fix that

Reply:
since it's the same file, simply not adding it to the archive again is actually nicer to avoid having the same data multiple times

Reply:
was thinking the revision info could be useful - maybe if you wanted know which files are from which revision, or maybe see only files from the latest revision. but yeah, this'll duplicate the files so if there are some large files with lots of revisions the export could be needlessly large..

Reply:
I'd expect people who want to do stuff with the data to process the yaml files and get the proper paths from there

Reply:
That could work if I update the export schema to include editing revisions in the yaml file (currently I just list all the files together)